File: monitoring-system/server/app/services/submission_service.py
"""
Submission service — CRUD + status management.
"""
from ..extensions import db
from ..models.submission import Submission
from ..models.user import User
from ..models.assignment import Assignment, AuditLog
from .notification_service import notify_user
from flask import request
from datetime import datetime
from sqlalchemy import or_
from ..models.status_history import SubmissionStatusHistory

STATUS_TRANSITIONS = {
    "pending": {"under_review", "assigned"},
    "under_review": {"assigned", "resolved"},
    "assigned": {"under_review", "resolved"},
    "resolved": {"closed"},
    "closed": set(),
}


def _audit(user_id, action, entity_id=None, details=None):
    log = AuditLog(user_id=user_id, action=action,
                   entity_type="submission", entity_id=entity_id,
                   details=details, ip_address=request.remote_addr)
    db.session.add(log)


def create_submission(user_id: int, data: dict) -> tuple[dict, int]:
    sub = Submission(
        user_id=user_id,
        title=data["title"],
        description=data["description"],
        category=data.get("category", "other"),
        priority=data.get("priority", "medium"),
    )
    db.session.add(sub)
    db.session.flush()
    db.session.add(SubmissionStatusHistory(
        submission_id=sub.id, changed_by=user_id, from_status=None, to_status=sub.status
    ))
    _audit(user_id, "CREATE_SUBMISSION", sub.id, f"Title: {sub.title}")

    # Notify all admins
    admins = User.query.filter_by(role="admin", is_active=True).all()
    for admin in admins:
        notify_user(admin.id, sub.id,
                    f"New submission: '{sub.title}' from {sub.submitter.name if sub.submitter else 'user'}")

    db.session.commit()
    return {"message": "Submission created.", "submission": sub.to_dict()}, 201


def get_submissions(user: User, page: int = 1, per_page: int = 20,
                    status: str = None, category: str = None,
                    priority: str = None, search: str = None,
                    date_from: str = None, date_to: str = None,
                    assigned_to: int = None) -> tuple[dict, int]:
    query = Submission.query
    if user.role == "user":
        query = query.filter_by(user_id=user.id)
    elif user.role == "secretary":
        # Only assigned submissions
        assigned_ids = [a.submission_id for a in user.assignments.all()]
        query = query.filter(Submission.id.in_(assigned_ids))
    # admin sees all
    if status:
        query = query.filter_by(status=status)
    if category:
        query = query.filter_by(category=category)
    if priority:
        query = query.filter_by(priority=priority)
    if assigned_to:
        query = query.join(Assignment, Assignment.submission_id == Submission.id)
        query = query.filter(Assignment.assigned_to == assigned_to)
    if search:
        term = f"%{search.strip()}%"
        query = query.filter(or_(Submission.title.ilike(term), Submission.description.ilike(term)))
    if date_from:
        try:
            query = query.filter(Submission.created_at >= datetime.fromisoformat(date_from))
        except ValueError:
            return {"error": "Invalid 'from' date. Use ISO-8601 format."}, 400
    if date_to:
        try:
            query = query.filter(Submission.created_at <= datetime.fromisoformat(date_to))
        except ValueError:
            return {"error": "Invalid 'to' date. Use ISO-8601 format."}, 400

    if page < 1 or per_page < 1 or per_page > 100:
        return {"error": "page must be >= 1 and per_page must be between 1 and 100."}, 400

    # FIFO queue: older submissions are attended to first; id breaks timestamp ties.
    paginated = query.order_by(Submission.created_at.asc(), Submission.id.asc()).paginate(
        page=page, per_page=per_page, error_out=False
    )
    return {
        "submissions": [s.to_dict() for s in paginated.items],
        "total":   paginated.total,
        "pages":   paginated.pages,
        "page":    page,
        "per_page": per_page,
    }, 200


def get_submission(sub_id: int, user: User) -> tuple[dict, int]:
    sub = Submission.query.get_or_404(sub_id)
    # Access control
    if user.role == "user" and sub.user_id != user.id:
        return {"error": "Access denied."}, 403
    if user.role == "secretary":
        assigned = Assignment.query.filter_by(
            submission_id=sub_id, assigned_to=user.id).first()
        if not assigned:
            return {"error": "Access denied."}, 403
    return {"submission": sub.to_dict(include_responses=True)}, 200


def update_status(sub_id: int, new_status: str, actor: User) -> tuple[dict, int]:
    sub = Submission.query.get_or_404(sub_id)
    old_status = sub.status
    if actor.role == "secretary" and not Assignment.query.filter_by(
            submission_id=sub_id, assigned_to=actor.id).first():
        return {"error": "You can only update submissions assigned to you."}, 403
    if new_status == old_status:
        return {"error": "Submission is already in that status."}, 409
    if new_status not in STATUS_TRANSITIONS.get(old_status, set()):
        return {"error": f"Invalid status transition from '{old_status}' to '{new_status}'."}, 409
    sub.status = new_status
    db.session.add(SubmissionStatusHistory(submission_id=sub.id, changed_by=actor.id,
                                           from_status=old_status, to_status=new_status))
    _audit(actor.id, "UPDATE_STATUS", sub.id,
           f"{old_status} → {new_status}")
    # Notify submitter
    notify_user(sub.user_id, sub.id,
                f"Your submission '{sub.title}' status changed to '{new_status}'.")
    db.session.commit()
    return {"message": "Status updated.", "submission": sub.to_dict()}, 200


def delete_submission(sub_id: int, actor: User) -> tuple[dict, int]:
    sub = Submission.query.get_or_404(sub_id)
    _audit(actor.id, "DELETE_SUBMISSION", sub_id, f"Deleted: {sub.title}")
    db.session.delete(sub)
    db.session.commit()
    return {"message": "Submission deleted."}, 200


def assign_submission(data: dict, actor: User) -> tuple[dict, int]:
    sub_id   = data["submission_id"]
    sec_id   = data["assigned_to"]
    notes    = data.get("notes", "")

    sub = Submission.query.get_or_404(sub_id)
    secretary = User.query.get_or_404(sec_id)
    if secretary.role != "secretary" or not secretary.is_active:
        return {"error": "Target user is not a secretary."}, 400

    # Remove old assignment if exists
    old = Assignment.query.filter_by(submission_id=sub_id).first()
    if old:
        db.session.delete(old)

    assignment = Assignment(
        submission_id=sub_id,
        assigned_to=sec_id,
        assigned_by=actor.id,
        notes=notes,
    )
    old_status = sub.status
    sub.status = "assigned"
    db.session.add(SubmissionStatusHistory(submission_id=sub.id, changed_by=actor.id,
                                           from_status=old_status, to_status="assigned"))
    db.session.add(assignment)
    db.session.flush()
    _audit(actor.id, "ASSIGN_SUBMISSION", sub_id,
           f"Assigned to secretary {sec_id}")
    notify_user(sec_id, sub_id,
                f"You have been assigned submission: '{sub.title}'")
    notify_user(sub.user_id, sub_id,
                f"Your submission '{sub.title}' has been assigned for review.")
    db.session.commit()
    return {"message": "Submission assigned.", "assignment": assignment.to_dict()}, 200


def get_status_history(sub_id: int, user: User) -> tuple[dict, int]:
    sub = Submission.query.get_or_404(sub_id)
    if user.role == "user" and sub.user_id != user.id:
        return {"error": "Access denied."}, 403
    if user.role == "secretary" and not Assignment.query.filter_by(
            submission_id=sub_id, assigned_to=user.id).first():
        return {"error": "Access denied."}, 403
    return {"status_history": [item.to_dict() for item in sub.status_history]}, 200

File: monitoring-system/server/src/services/submissionService.js
import { prisma } from '../db.js';
import { HttpError } from '../utils/httpError.js';
import { parseIsoDate } from '../utils/dates.js';
import { submissionDict, assignmentDict, statusHistoryDict } from '../utils/serialize.js';
import { writeAudit, notifyUser } from './audit.js';

const STATUS_TRANSITIONS = {
  pending: ['under_review', 'assigned'],
  under_review: ['assigned', 'resolved'],
  assigned: ['under_review', 'resolved'],
  resolved: ['closed'],
  closed: [],
};

const SUB_WITH = {
  submitter: true,
  responses: { include: { responder: true }, orderBy: { created_at: 'asc' } },
  status_history: { include: { actor: true }, orderBy: { changed_at: 'asc' } },
};

async function findSubmissionOr404(id, include) {
  const sub = await prisma.submissions.findUnique({ where: { id }, include });
  if (!sub) throw new HttpError(404, 'Resource not found.');
  return sub;
}

export async function createSubmission(userId, data, ip) {
  const submitter = await prisma.users.findUnique({ where: { id: userId } });
  const created = await prisma.$transaction(async (tx) => {
    const sub = await tx.submissions.create({
      data: {
        user_id: userId,
        title: data.title,
        description: data.description,
        category: data.category ?? 'other',
        priority: data.priority ?? 'medium',
      },
    });
    await tx.submission_status_history.create({
      data: { submission_id: sub.id, changed_by: userId, from_status: null, to_status: sub.status },
    });
    await writeAudit(tx, {
      userId, action: 'CREATE_SUBMISSION', entityType: 'submission', entityId: sub.id,
      details: `Title: ${sub.title}`, ip,
    });
    const admins = await tx.users.findMany({ where: { role: 'admin', is_active: true } });
    for (const admin of admins) {
      await notifyUser(tx, admin.id, sub.id,
        `New submission: '${sub.title}' from ${submitter ? submitter.name : 'user'}`);
    }
    return sub;
  });
  const full = await findSubmissionOr404(created.id, { submitter: true });
  return { status: 201, body: { message: 'Submission created.', submission: submissionDict(full) } };
}

export async function getSubmissions(user, {
  page = 1, perPage = 20, status, category, priority, search, from, to, assignedTo,
} = {}) {
  const where = {};
  if (user.role === 'user') {
    where.user_id = user.id;
  } else if (user.role === 'secretary') {
    const mine = await prisma.assignments.findMany({
      where: { assigned_to: user.id }, select: { submission_id: true },
    });
    where.id = { in: mine.map((a) => a.submission_id) };
  }
  if (status) where.status = status;
  if (category) where.category = category;
  if (priority) where.priority = priority;
  if (assignedTo) where.assignment = { assigned_to: assignedTo };
  if (search) {
    const term = search.trim();
    where.OR = [{ title: { contains: term } }, { description: { contains: term } }];
  }
  if (from != null) {
    const d = parseIsoDate(from);
    if (!d) throw new HttpError(400, "Invalid 'from' date. Use ISO-8601 format.");
    where.created_at = { ...(where.created_at || {}), gte: d };
  }
  if (to != null) {
    const d = parseIsoDate(to);
    if (!d) throw new HttpError(400, "Invalid 'to' date. Use ISO-8601 format.");
    where.created_at = { ...(where.created_at || {}), lte: d };
  }

  if (page < 1 || perPage < 1 || perPage > 100) {
    throw new HttpError(400, 'page must be >= 1 and per_page must be between 1 and 100.');
  }

  const [total, items] = await prisma.$transaction([
    prisma.submissions.count({ where }),
    prisma.submissions.findMany({
      where,
      include: { submitter: true },
      orderBy: [{ created_at: 'asc' }, { id: 'asc' }],
      skip: (page - 1) * perPage,
      take: perPage,
    }),
  ]);

  return {
    submissions: items.map((s) => submissionDict(s)),
    total,
    pages: total === 0 ? 0 : Math.ceil(total / perPage),
    page,
    per_page: perPage,
  };
}

async function assertCanView(sub, user) {
  if (user.role === 'user' && sub.user_id !== user.id) {
    throw new HttpError(403, 'Access denied.');
  }
  if (user.role === 'secretary') {
    const a = await prisma.assignments.findFirst({
      where: { submission_id: sub.id, assigned_to: user.id },
    });
    if (!a) throw new HttpError(403, 'Access denied.');
  }
}

export async function getSubmission(subId, user) {
  const sub = await findSubmissionOr404(subId, SUB_WITH);
  await assertCanView(sub, user);
  return { submission: submissionDict(sub, { includeResponses: true }) };
}

export async function getStatusHistory(subId, user) {
  const sub = await findSubmissionOr404(subId, {
    status_history: { include: { actor: true }, orderBy: { changed_at: 'asc' } },
  });
  await assertCanView(sub, user);
  return { status_history: sub.status_history.map(statusHistoryDict) };
}

export async function updateStatus(subId, newStatus, actor, ip) {
  const sub = await findSubmissionOr404(subId);
  const oldStatus = sub.status;
  if (actor.role === 'secretary') {
    const a = await prisma.assignments.findFirst({
      where: { submission_id: subId, assigned_to: actor.id },
    });
    if (!a) throw new HttpError(403, 'You can only update submissions assigned to you.');
  }
  if (newStatus === oldStatus) {
    throw new HttpError(409, 'Submission is already in that status.');
  }
  if (!(STATUS_TRANSITIONS[oldStatus] || []).includes(newStatus)) {
    throw new HttpError(409, `Invalid status transition from '${oldStatus}' to '${newStatus}'.`);
  }
  await prisma.$transaction(async (tx) => {
    await tx.submissions.update({ where: { id: subId }, data: { status: newStatus } });
    await tx.submission_status_history.create({
      data: { submission_id: subId, changed_by: actor.id, from_status: oldStatus, to_status: newStatus },
    });
    await writeAudit(tx, {
      userId: actor.id, action: 'UPDATE_STATUS', entityType: 'submission', entityId: subId,
      details: `${oldStatus} → ${newStatus}`, ip,
    });
    await notifyUser(tx, sub.user_id, subId,
      `Your submission '${sub.title}' status changed to '${newStatus}'.`);
  });
  const full = await findSubmissionOr404(subId, { submitter: true });
  return { message: 'Status updated.', submission: submissionDict(full) };
}

export async function deleteSubmission(subId, actor, ip) {
  const sub = await findSubmissionOr404(subId);
  await prisma.$transaction(async (tx) => {
    await writeAudit(tx, {
      userId: actor.id, action: 'DELETE_SUBMISSION', entityType: 'submission', entityId: subId,
      details: `Deleted: ${sub.title}`, ip,
    });
    await tx.responses.deleteMany({ where: { submission_id: subId } });
    await tx.submission_status_history.deleteMany({ where: { submission_id: subId } });
    await tx.notifications.deleteMany({ where: { submission_id: subId } });
    await tx.assignments.deleteMany({ where: { submission_id: subId } });
    await tx.submissions.delete({ where: { id: subId } });
  });
  return { message: 'Submission deleted.' };
}

export async function assignSubmission({ submission_id, assigned_to, notes = '' }, actor, ip) {
  const sub = await findSubmissionOr404(submission_id);
  const secretary = await prisma.users.findUnique({ where: { id: assigned_to } });
  if (!secretary) throw new HttpError(404, 'Resource not found.');
  if (secretary.role !== 'secretary' || !secretary.is_active) {
    throw new HttpError(400, 'Target user is not a secretary.');
  }
  const assignment = await prisma.$transaction(async (tx) => {
    await tx.assignments.deleteMany({ where: { submission_id } });
    const a = await tx.assignments.create({
      data: { submission_id, assigned_to, assigned_by: actor.id, notes },
    });
    await tx.submissions.update({ where: { id: submission_id }, data: { status: 'assigned' } });
    await tx.submission_status_history.create({
      data: { submission_id, changed_by: actor.id, from_status: sub.status, to_status: 'assigned' },
    });
    await writeAudit(tx, {
      userId: actor.id, action: 'ASSIGN_SUBMISSION', entityType: 'submission', entityId: submission_id,
      details: `Assigned to secretary ${assigned_to}`, ip,
    });
    await notifyUser(tx, assigned_to, submission_id, `You have been assigned submission: '${sub.title}'`);
    await notifyUser(tx, sub.user_id, submission_id,
      `Your submission '${sub.title}' has been assigned for review.`);
    return a;
  });
  const full = await prisma.assignments.findUnique({
    where: { id: assignment.id }, include: { secretary: true, admin: true },
  });
  return { message: 'Submission assigned.', assignment: assignmentDict(full) };
}

export async function listAssignments(user) {
  const where = user.role === 'secretary' ? { assigned_to: user.id } : {};
  const items = await prisma.assignments.findMany({
    where, include: { secretary: true, admin: true }, orderBy: { id: 'asc' },
  });
  return { assignments: items.map(assignmentDict) };
}
